### PR TITLE
Feature: captcha for Reset Password & Resend Confirmation

### PR DESCRIPTION
**Pull request by josadcha**

### CODE REVIEW

File: src/screens/EmailVerification/index.tsx [hunk 1]
```diff
@@ -24,63 +29,183 @@ interface OwnProps {
...
     emailVerificationFetch,
 };
 
-//tslint:disable-next-line:no-any
-export const EmailVerificationScreen = injectIntl(withRouter(connect(mapStateToProps, mapDispatchProps)(EmailVerificationComponent) as any));
+export const EmailVerificationScreen = compose(
+    injectIntl,
+    withRouter,
+    connect(mapStateToProps, mapDispatchProps),
+)(EmailVerificationComponent) as any; // tslint:disable-line
```
Comment (akhlopiachyi) [inline src/screens/EmailVerification/index.tsx:211]: replace 'as any' with 'as React.ComponentType'

File: src/screens/ForgotPassword/index.tsx [hunk 2]
```diff
@@ -128,20 +218,40 @@ class ForgotPasswordComponent extends React.Component<Props, ForgotPasswordState
...
         forgotPassword: credentials => dispatch(forgotPassword(credentials)),
     });
 
-// tslint:disable-next-line:no-any
-export const ForgotPasswordScreen = injectIntl(withRouter(connect(mapStateToProps, mapDispatchProps)(ForgotPasswordComponent) as any));
+export const ForgotPasswordScreen = compose(
+    injectIntl,
+    withRouter,
+    connect(mapStateToProps, mapDispatchToProps),
+)(ForgotPasswordComponent) as any; // tslint:disable-line
```
Comment (akhlopiachyi) [inline src/screens/ForgotPassword/index.tsx:257]: the same as above
Also add an empty line in the end of file

File: src/components/EmailForm/index.tsx [hunk 3]
```diff
@@ -35,14 +40,20 @@ export class EmailForm extends React.Component<EmailFormProps> {
             email,
             emailFocused,
             emailError,
+            captchaType,
         } = this.props;
+
+        const emailFormClass = cr('cr-email-form', {
+            'cr-email-form-high': captchaType && captchaType !== 'none',
```
Comment (oyershov) [inline src/components/EmailForm/index.tsx:47]:
```suggestion
            'cr-email-form--high': captchaType && captchaType !== 'none',
```

File: src/components/EmailForm/index.tsx [hunk 4]
```diff
@@ -72,11 +83,12 @@ export class EmailForm extends React.Component<EmailFormProps> {
                             />
                             {emailError && <div className="cr-email-form__error">{emailError}</div>}
                         </div>
+                        {this.props.renderCaptcha}
                         <div className="cr-email-form__button-wrapper">
                             <Button
                                 block={true}
                                 type="button"
-                                disabled={isLoading || !email.match(EMAIL_REGEX)}
+                                disabled={this.disableButton()}
```
Comment (oyershov) [inline src/components/EmailForm/index.tsx:91]:
```suggestion
                                disabled={this.isButtonDisabled()}
```

File: src/screens/ForgotPassword/index.tsx [hunk 5]
```diff
@@ -35,38 +44,95 @@ interface ForgotPasswordState {
...
 
+    private reCaptchaRef;
+    private geetestCaptchaRef;
+
     public componentDidMount() {
         setDocumentTitle('Forgot password');
     }
 
+    public UNSAFE_componentWillReceiveProps(nextProps: Props) {
+        if (nextProps.error) {
```
Comment (oyershov) [inline src/screens/ForgotPassword/index.tsx:80]:
```suggestion
        if (nextProps.error && !this.props.error) {
```

**Comment by pkucherenk0** [comment]
1. After success or failure resetting the password or confirmation code the checkbox of Recaptcha have to be reset. 
2. On resend confirmation email we should not be able to resend second time without passing new Recaptcha
[vokoscreen-2020-05-21_13-49-52.zip](https://github.com/openware/baseapp/files/4662050/vokoscreen-2020-05-21_13-49-52.zip)
3. On resend confirmation email geetest looks too wide 
Selection_999(245)

**Comment by josadcha** [comment]
Needs 1-2h to finish

**Comment by josadcha** [comment]
Screenshot 2020-09-07 at 19 03 10
fixed styles + add captcha reset in case of error or success

File: src/components/EmailForm/index.tsx [hunk 6]
```diff
@@ -65,14 +89,17 @@ const EmailForm = React.memo((props: EmailFormProps) => {
         }
     };
 
+    const emailFormClass = cr('cr-email-form', {
+        'cr-email-form--high': captchaType && captchaType !== 'none',
```
Comment (oyershov) [inline src/components/EmailForm/index.tsx:93]:
```suggestion
        'cr-email-form--extended': captchaType && captchaType !== 'none',
```

File: src/screens/EmailVerification/index.tsx [hunk 7]
```diff
@@ -32,18 +48,80 @@ interface DispatchProps {
...
+    public translate = (id: string) => this.props.intl.formatMessage({ id });
+
+    public renderCaptcha = () => {
+        const { shouldGeetestReset } = this.state;
+        const { configs } = this.props;
+
+        switch (configs.captcha_type) {
+            case 'recaptcha':
+                return (
+                    <div className="pg-emailverification-recaptcha">
```
Comment (oyershov) [inline src/screens/EmailVerification/index.tsx:102]:
```suggestion
                    <div className="pg-emailverification--recaptcha">
```

File: src/screens/EmailVerification/index.tsx [hunk 8]
```diff
@@ -32,18 +48,80 @@ interface DispatchProps {
...
+                        <ReCAPTCHA
+                            ref={this.reCaptchaRef}
+                            sitekey={configs.captcha_id}
+                            onChange={this.handleReCaptchaSuccess}
+                        />
+                    </div>
+                );
+            case 'geetest':
+                return (
+                    <div className="pg-emailverification-geetest">
```
Comment (oyershov) [inline src/screens/EmailVerification/index.tsx:112]:
```suggestion
                    <div className="pg-emailverification--geetest">
```

**Review by oyershov — CHANGES_REQUESTED** [review]
Please, rework this PR.
1. Define separated component <Captha /> with all logic related to the captcha verification
2. Remove all logic related captcha from other components ```EmailVerification```, ```ForgotPassword```, ```SignUpScreen```

**Review by berezh — APPROVED** [review]

File: src/screens/EmailVerification/index.tsx [hunk 9]
```diff
@@ -23,6 +29,15 @@ interface OwnProps {
...
+    success: boolean;
+    error?: CommonError;
+}
+
+interface VerificationState {
+    captcha_response: string;
+    reCaptchaSuccess: boolean;
+    geetestCaptchaSuccess: boolean;
+    shouldGeetestReset: boolean;
 }
```
Comment (oyershov) [inline src/screens/EmailVerification/index.tsx:41]: move this state to the ```Captcha``` component

File: src/screens/EmailVerification/index.tsx [hunk 10]
```diff
@@ -84,18 +131,76 @@ class EmailVerificationComponent extends React.Component<Props> {
...
+        });
+    };
+
+    private handleGeetestCaptchaSuccess = value => {
+        this.setState({
+            geetestCaptchaSuccess: true,
+            captcha_response: value,
+            shouldGeetestReset: false,
+        });
+    };
```
Comment (oyershov) [inline src/screens/EmailVerification/index.tsx:192]: move this code to the ```captcha``` component too

File: src/screens/ForgotPassword/index.tsx [hunk 11]
```diff
@@ -36,6 +42,10 @@ interface ForgotPasswordState {
     email: string;
     emailError: string;
     emailFocused: boolean;
+    captcha_response: string;
+    reCaptchaSuccess: boolean;
+    geetestCaptchaSuccess: boolean;
+    shouldGeetestReset: boolean;
```
Comment (oyershov) [inline src/screens/ForgotPassword/index.tsx:48]: Same as above.

File: src/screens/ForgotPassword/index.tsx [hunk 12]
```diff
@@ -127,14 +182,31 @@ class ForgotPasswordComponent extends React.Component<Props, ForgotPasswordState
...
+    };
+
+    private handleGeetestCaptchaSuccess = value => {
+        this.setState({
+            geetestCaptchaSuccess: true,
+            captcha_response: value,
+            shouldGeetestReset: false,
+        });
+    };
 }
```
Comment (oyershov) [inline src/screens/ForgotPassword/index.tsx:203]: This is an example of duplicating code

**Review by oyershov — CHANGES_REQUESTED** [review]

**Comment by pkucherenk0** [comment]
1. On the Forgot password page only the first password reset attempt will renew the captcha/geetest, all following attempts will force the user to wait for a minute until the captcha/geetest, will expire by time.

**Review by akhlopiachyi — APPROVED** [review]

**Pull request merged**